src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -225,31 +285,158 @@ private TableEntity lookupByFilePath(
-      if (!entities.iterator().hasNext()) {
+      int count =
+          TableServiceClientUtils.getTableEntryCount(tableServiceClient, tableName, options);
+      if (count < 1) {
         return null;
+      } else if (count > 1) {
+        throw new FileSystemAbortTransactionException(
+            String.format("lookupByFileId found too many entries for fileId %s", fileId));
       }
-      int count = 0;
-      for (TableEntity entity : entities) {
-        count += 1;
-        if (count > 1) {
-          logger.warn("Found more than one entry for file {}", fileId);
-          throw new FileSystemAbortTransactionException("lookupByFileId found too many entries");
-        }
-      }
+      TableClient tableClient = tableServiceClient.getTableClient(tableName);
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
       return entities.iterator().next();
 
     } catch (TableServiceException ex) {
       throw new FileSystemExecutionException("lookupByFileId operation failed");
     }
   }
 
-  // TODO: Implement snapshot-specific methods
+  // Returns empty list if not found
+  private List<TableEntity> batchLookupByFileId(
+      TableServiceClient tableServiceClient, List<String> fileIds) {
+    return ListUtils.partition(fileIds, MAX_FILTER_CLAUSES).stream()
+        .flatMap(
+            fileIdsBatch -> {
+              List<TableEntity> entities =
+                  TableServiceClientUtils.batchRetrieveFiles(tableServiceClient, fileIdsBatch);
+              return entities.stream();
+            })
+        .collect(Collectors.toList());
+  }
+
+  // -- Snapshot filesystem methods --
 
+  // To improve performance of building the snapshot file system, we use three techniques:
+  // 1. Operate over batches so that we can issue requests to fire store in parallel
+  // 2. Cache directory paths so that we do not due extra lookups or creates on shared directory
+  // structure
+  // 3. Rewrite rather than read, check existence, and then write. The logic here is that there is
+  // no contention
+  //    so the writing doesn't generate conflicts, and the typical use case is that overwrites will
+  // be rare:
+  //      a. File references are usually unique in the datasets we know about
+  //      b. Directories are cached, so will be overwritten based on the effectiveness of the cache
+
+  // Q - Do we need to support different table service clients for snapshots?? (like the dataset vs.
+  // snapshot firestore?)
+  public void addEntriesToSnapshot(
+      TableServiceClient datasetTableServiceClient,
+      TableServiceClient snapshotTableServiceClient,
+      String datasetId,
+      String datasetDirName,
+      String snapshotId,
+      List<String> fileIds) {
+    LRUMap<String, Boolean> pathMap = new LRUMap<>(MAX_FILTER_CLAUSES);
+    storeTopDirectory(snapshotTableServiceClient, snapshotId, datasetDirName);
+    ListUtils.partition(fileIds, MAX_FILTER_CLAUSES).stream()
+        .forEach(
```

Comment:
There's no need to construct a `Stream` here. You can just do `ListUtils.partition(...).forEach(...), which will do a simple for-loop underneath.